src/CloudNative.CloudEvents.AspNetCore/HttpRequestExtension.cs

```diff
@@ -45,90 +22,54 @@ public static class HttpRequestExtension
         /// <param name="extensions">List of extension instances</param>
         /// <returns>A CloudEvent instance or 'null' if the request message doesn't hold a CloudEvent</returns>
         public static async ValueTask<CloudEvent> ReadCloudEventAsync(this HttpRequest httpRequest,
-            ICloudEventFormatter formatter = null,
-            params ICloudEventExtension[] extensions)
+            ICloudEventFormatter formatter,
+            params CloudEventAttribute[] extensionAttributes)
         {
-            if (httpRequest.ContentType != null &&
-                httpRequest.ContentType.StartsWith(CloudEvent.MediaType,
-                    StringComparison.InvariantCultureIgnoreCase))
+            if (HasCloudEventsContentType(httpRequest))
             {
-                // handle structured mode
-                if (formatter == null)
-                {
-                    // if we didn't get a formatter, pick one
-                    var contentType = httpRequest.ContentType.Split(';')[0].Trim();
-                    if (contentType.EndsWith(JsonEventFormatter.MediaTypeSuffix,
-                        StringComparison.InvariantCultureIgnoreCase))
-                    {
-                        formatter = jsonFormatter;
-                    }
-                    else
-                    {
-                        throw new InvalidOperationException("Unsupported CloudEvents encoding");
-                    }
-                }
-
-                return await formatter.DecodeStructuredEventAsync(httpRequest.Body, extensions);
+                // TODO: Handle formatter being null
+                return await formatter.DecodeStructuredEventAsync(httpRequest.Body, extensionAttributes).ConfigureAwait(false);
             }
             else
             {
+                var headers = httpRequest.Headers;
                 CloudEventsSpecVersion version = CloudEventsSpecVersion.Default;
-                if (httpRequest.Headers[SpecVersionHttpHeader1] != StringValues.Empty)
+                if (headers.TryGetValue(HttpUtilities.SpecVersionHttpHeader, out var values))
                 {
-                    version = CloudEventsSpecVersion.V0_1;
+                    string versionId = values.First();
+                    version = CloudEventsSpecVersion.FromVersionId(versionId);
                 }
 
-                if (httpRequest.Headers[SpecVersionHttpHeader2] != StringValues.Empty)
+                var cloudEvent = new CloudEvent(version, extensionAttributes);
+                foreach (var header in headers)
                 {
-                    switch (httpRequest.Headers[SpecVersionHttpHeader2])
+                    string attributeName = HttpUtilities.GetAttributeNameFromHeaderName(header.Key);
+                    if (attributeName is null || attributeName == CloudEventsSpecVersion.SpecVersionAttribute.Name)

```

Comment:
So if ce-specversion isn't present, but specversion is we would just ignore it?

Reply:
Yes, absolutely - we should ignore all headers that don't start with "ce-". (This isn't a change - although what *has* changed is that we don't use "ce-cloudeventsversion" which was a 0.1/0.2 header.)